src/org/rascalmpl/repl/TerminalProgressBarMonitor.java

```diff
@@ -0,0 +1,459 @@
+            pb.max += totalWork;
+            pb.nesting++;
+            pb.update();
+        }
+    }
+
+    @Override
+    public synchronized void jobStep(String name, String message, int workShare) {
+        ProgressBar pb = findBarByName(name);
+        
+        if (pb != null) {
+            pb.worked(workShare, message);
+            pb.update();
+        }
+    }
+
+    @Override
+    public synchronized int jobEnd(String name, boolean succeeded) {
+        var pb = findBarByName(name);
+
+        if (pb != null && --pb.nesting == -1) {
+            eraseBars();
+            // write it one last time into the scrollback buffer (on top)
+            pb.done();
+            pb.write();
+            bars.remove(pb);
+            // print the left over bars under this one.
+            printBars();
+            return pb.current;
+        }
+        else if (pb != null) {
+            pb.done();
+            pb.update();
+        }
+
+        return -1;
+    }
+
+    @Override
+    public synchronized boolean jobIsCanceled(String name) {
+       // ? don't know what this should do
+       return false;
+    }
+
+    @Override
+    public synchronized void jobTodo(String name, int work) {
+        ProgressBar pb = findBarByName(name);
+        
+        if (pb != null) {
+            pb.max += work;
+            pb.update();
+        }
+    }
+
+    @Override
+    public synchronized void warning(String message, ISourceLocation src) {
+        if (bars.size() > 0) {
+            eraseBars();
+        }
+        writer.println(("[WARNING] " + src + ": " + message));
+        if (bars.size() > 0) {
+            printBars();
+        }
+    }
+
+    /**
+     * Here we make sure the progress bars are gone just before
+     * someone wants to print in the console. When the printing
+     * is ready, we simply add our own progress bars again.
+     */
+    @Override
+    public synchronized void write(byte[] b) throws IOException {
+        if (bars.size() > 0) {
+            eraseBars();
+            out.write(b);
+            printBars();
+        }
+        else {
+            out.write(b);
+        }
+    }
+
+    /**
+     * Here we make sure the progress bars are gone just before
+     * someone wants to print in the console. When the printing
+     * is ready, we simply add our own progress bars again.
+     */
+    @Override
+    public synchronized void write(byte[] b, int off, int len) throws IOException {
+    if (bars.size() > 0) {
+            eraseBars();
+            out.write(b, off, len);
+            printBars();
+        }
+        else {
+            out.write(b, off, len);
+        }
+    }
+
+    /**
+     * Here we make sure the progress bars are gone just before
+     * someone wants to print in the console. When the printing
+     * is ready, we simply add our own progress bars again.
+     */
+    @Override
+    public synchronized void write(int b) throws IOException {
```

Comment:
shouldn't we also override `flush`?

Reply:
no reason. `write` captures everything that comes in that we need to wrap with ANSI commands to remove the bars  (before) and the printing of the new bars (after). Flushing is only necessary when the context needs a flush, or during debugging to be able to see the cursor move to the ANSI commands.